src/main/flight/position.c

```diff
@@ -137,16 +142,40 @@ void calculateEstimatedAltitude(timeUs_t currentTimeUs)
     }
 #endif
 
-    if (ARMING_FLAG(ARMED) && !altitudeOffsetSet) {
+    if (ARMING_FLAG(ARMED) && !altitudeOffsetSetBaro) {
         baroAltOffset = baroAlt;
-        gpsAltOffset = gpsAlt;
-        altitudeOffsetSet = true;
-    } else if (!ARMING_FLAG(ARMED) && altitudeOffsetSet) {
-        altitudeOffsetSet = false;
+        altitudeOffsetSetBaro = true;
+    } else if (!ARMING_FLAG(ARMED) && altitudeOffsetSetBaro) {
+        altitudeOffsetSetBaro = false;
     }
+
     baroAlt -= baroAltOffset;
+
+    int goodGpsSats = 0;
+    int badGpsSats = -1;
+
+    if (haveBaroAlt) {
+        goodGpsSats = positionConfig()->altNumSatsGpsUse;
+        badGpsSats = positionConfig()->altNumSatsBaroFallback;
+    }
+
+    if (ARMING_FLAG(ARMED)) {
+        if (!altitudeOffsetSetGPS && gpsNumSat >= goodGpsSats) {
```

Comment:
This code is a suboptimal - gpsAltOffset is absolute value, Once you get good value, the offset is valid until disarmed (until flying on place with different altitude).  Also BARO will drift, so new offset on second fix is less accurate than first one.
Yet another problem is that noisy data are sampled here, making ofset error larger than necessary.


Kalman will help here a lot - modelling baro offset and them mixing baro/gps ...

Reply:
Yeah, it's sub-optimal, but I don't plan to make a 100% robust implementation here. It's a single specific bug fix. Actually I planned to work on a good and robust altitude estimation later - it's a separate big task including baro / errors handling / sensors smoothing / etc.